chapter3/src/order.go

```diff
@@ -0,0 +1,162 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"time"
+)
+
+type Order struct {
+	Id          string
+	orderLines  []OrderLine
+	total       float32
+	dateCreated time.Time
+	user        string
+}
+
+type OrderLine struct {
+	Id          string
+	item        string
+	dateCreated time.Time
+	quantity    int16
+	unitPrice   float32
+}
+
+func CreateOrder(id string, orderLine []OrderLine, dateCreated time.Time, user string) (*Order, error) {
+
+	var total float32 = 0
+
+	if id == "" {
+		return nil, errors.New("Some values are missing")
+	} else if orderLine == nil {
+		return nil, errors.New("Some values are missing")
+	} else if dateCreated.IsZero() {
+		return nil, errors.New("Some values are missing")
+	} else if user == "" {
+		return nil, errors.New("Some values are missing")
+	} else {
+		total = calculateCost(orderLine)
+	}
+
+	return &Order{
+		Id:          id,
+		orderLines:  orderLine,
+		total:       total,
+		dateCreated: dateCreated,
+		user:        user,
+	}, nil
+
+}
+
+func AppendOrderLine(orderLine OrderLine, order *Order) {
```

Comment:
You can make functions part of the object like this: 

`func (order *Order) AppendOrderLine(orderLine OrderLine)`

Then you can call this function like this: 

`order.AppendOrderLine(orderLine)`